### DISCUSSION TITLE
Profile Algorithm

### DISCUSSION THREAD

**Discussion by chanbinna** [UCSD-Likelion/Roomiverse#8] [General]
# 🏠 **C# .NET Core-Based Roommate & Housing Recommendation Algorithm**
1️⃣ On Campus / Off Campus filter -> hard filtering
2️⃣ for On Campus hard filter for the college -> hard filtering
3️⃣ Hard filtered for the required section (e.g. smoking, pet, gender)
4️⃣ for other feature vector -> use cosine similarity to solve it

Under is the ChatGPT explanation. 

## 🔥 **Concept**
---
This algorithm is designed to **analyze user data** and recommend the best roommates and housing options.  
The core components include:

1. **Hard Filtering (Mandatory Conditions)**: Criteria that must match (e.g., smoking status, gender, pets, etc.).
2. **Soft Filtering (Preference Scoring)**: Matching scores based on lifestyle preferences.
3. **Housing Recommendation Algorithm**: Taking distance, price range, and room type into account.
4. **Similar User Matching**: Applying cosine similarity or distance-based metrics.
5. **Weight Adjustment**: Customizing scores based on user priorities.

---
## 📌 **Architecture**
---
### **1️⃣ Tech Stack**
- **Backend**: C# .NET Core (ASP.NET Core Web API)
- **Database**: SQL Server or PostgreSQL (using Entity Framework Core)
- **Recommendation Algorithm**: Cosine Similarity + KNN (K-Nearest Neighbors)
- **Caching**: Redis (for performance optimization)

---
### **2️⃣ Database Schema**
```sql
CREATE TABLE Users (
    UserId INT PRIMARY KEY IDENTITY(1,1),
    Email NVARCHAR(255) UNIQUE NOT NULL,
    Name NVARCHAR(100),
    Gender NVARCHAR(10),
    BirthDate DATE,
    PasswordHash NVARCHAR(255),
    Instagram NVARCHAR(255) NULL,
    Facebook NVARCHAR(255) NULL,
    IsOffCampus BIT,
    GuestFrequency INT,
    PreferSameGender BIT,
    PetPreference NVARCHAR(50),
    SmokingStatus NVARCHAR(50),
    OkWithSmokers BIT,
    DrinkingStatus BIT,
    OkWithDrinkers BIT,
    SleepTime TIME,
    WakeTime TIME,
    SleepImportance INT,
    CleaningFrequency INT,
    Major NVARCHAR(100),
    College NVARCHAR(100),
    Ethnicity NVARCHAR(50)
);

CREATE TABLE HousingPreferences (
    UserId INT PRIMARY KEY,
    DistanceFromSchool FLOAT,
    PreferredPriceRange NVARCHAR(50),
    RoomType NVARCHAR(50),
    FOREIGN KEY (UserId) REFERENCES Users(UserId)
);
```

---
## 🎯 **Recommendation Algorithm Implementation**
---

### **1️⃣ Hard Filtering (Mandatory Conditions)**
First, filter users based on conditions that **must match**.
```csharp
var filteredUsers = users.Where(u =>
    u.IsOffCampus == targetUser.IsOffCampus &&
    (u.PreferSameGender == false || u.Gender == targetUser.Gender) &&
    (u.OkWithSmokers || u.SmokingStatus == "non" || targetUser.SmokingStatus == "non") &&
    (u.OkWithDrinkers || u.DrinkingStatus == false || targetUser.DrinkingStatus == false) &&
    (u.PetPreference == "No Preference" || u.PetPreference == targetUser.PetPreference)
).ToList();
```
At this stage, **basic roommate filtering** is applied.

---
### **2️⃣ Similarity-Based Recommendation (Cosine Similarity)**
Convert user preference data into vectors and use **cosine similarity** to calculate matching scores.

#### **✅ Vector Representation**
Convert user data into **numerical vectors** for similarity measurement.
```csharp
public double[] ConvertToVector(User user)
{
    return new double[]
    {
        user.GuestFrequency,
        user.SleepTime.Hour + user.SleepTime.Minute / 60.0,
        user.WakeTime.Hour + user.WakeTime.Minute / 60.0,
        user.SleepImportance,
        user.CleaningFrequency
    };
}
```

#### **✅ Cosine Similarity Calculation**
```csharp
public double CosineSimilarity(double[] vec1, double[] vec2)
{
    double dotProduct = vec1.Zip(vec2, (a, b) => a * b).Sum();
    double magnitude1 = Math.Sqrt(vec1.Sum(x => x * x));
    double magnitude2 = Math.Sqrt(vec2.Sum(x => x * x));

    return dotProduct / (magnitude1 * magnitude2);
}
```

#### **✅ Finding Similar Roommates**
```csharp
var targetVector = ConvertToVector(targetUser);
var rankedUsers = filteredUsers
    .Select(u => new { User = u, Score = CosineSimilarity(targetVector, ConvertToVector(u)) })
    .OrderByDescending(x => x.Score)
    .Take(5) // Recommend top 5 roommates
    .ToList();
```

---
### **3️⃣ Housing Recommendation Algorithm**
The housing recommendation considers **distance, price range, and room type** to calculate scores.

#### **✅ Distance Score Calculation**
```csharp
double CalculateDistanceScore(double userDistance, double targetDistance)
{
    return 1 / (1 + Math.Abs(userDistance - targetDistance));
}
```

#### **✅ Price Score Calculation**
```csharp
int CalculatePriceScore(string userPrice, string targetPrice)
{
    return userPrice == targetPrice ? 10 : 5; // Higher score if in the same price range
}
```

#### **✅ Final Score Calculation**
```csharp
var rankedHousing = housingList
    .Select(h => new {
        Housing = h,
        Score = CalculateDistanceScore(h.DistanceFromSchool, targetHousing.DistanceFromSchool) * 0.6 +
                CalculatePriceScore(h.PreferredPriceRange, targetHousing.PreferredPriceRange) * 0.4
    })
    .OrderByDescending(x => x.Score)
    .Take(5) // Recommend top 5 housing options
    .ToList();
```

---
## **🎉 Final Implementation Flow**
1. **User Login**
2. **Hard Filtering** (Check mandatory conditions)
3. **Find Similar Users** (Apply cosine similarity)
4. **Recommend Housing** (Score based on distance, price, and room type)
5. **Return Recommendations**

---
## 🚀 **Example Output**
```json
{
    "roommate_recommendations": [
        { "name": "John Doe", "similarity_score": 0.85 },
        { "name": "Alice Kim", "similarity_score": 0.78 },
        { "name": "David Lee", "similarity_score": 0.72 }
    ],
    "housing_recommendations": [
        { "location": "123 Main St", "score": 9.5 },
        { "location": "456 Oak St", "score": 8.7 },
        { "location": "789 Pine St", "score": 8.2 }
    ]
}
```

---
## **💡 Optimization Strategies**
- **Adjust Similarity Weights**: Allow users to prioritize certain factors.
- **Use KNN**: Implement K-Nearest Neighbors to find the most similar users.
- **Redis Caching**: Cache similarity results for faster lookups.

---
## ✅ **Conclusion**
This system is implemented as a **.NET Core-based RESTful API** and utilizes a **cosine similarity-based recommendation system** to match the best roommates and housing options.  
The key approach is **hard filtering → cosine similarity → scoring based on distance and price**.

Now, you can proceed with implementing the API using this architecture! 🚀

**Comment by 7174Andy** [comment]
온캠 / 오프캠 filter -> College filter -> hard-filters on other features (e.g. smoke, pet, etc) -> make other features vector -> cosine similarity

**Comment by chanbinna** [comment]
<html><head></head><body><h3><strong>TorchSharp API Limitations Compared to Python PyTorch</strong></h3>
<p>TorchSharp is the official .NET binding for PyTorch, but it <strong>does not have full parity</strong> with Python’s PyTorch. Here are some key limitations:</p>
<hr>
<h2><strong>🚨 Missing or Limited Features in TorchSharp</strong></h2>

Feature | TorchSharp Support | Notes
-- | -- | --
Deep Learning Models | 🚫 Limited | No torchvision.models, no direct torch.nn.Sequential equivalent
Pretrained Models | 🚫 No | Cannot load PyTorch pretrained models (torch.hub.load())
TorchScript & ONNX | 🚫 No | No torch.jit.script() or torch.onnx.export() support
Custom CUDA Kernels | 🚫 No | Cannot define custom CUDA operations
Autograd & Backpropagation | ✅ Yes | Supports basic automatic differentiation
Tensor Operations | ✅ Partial | Most tensor functions work, but some may be missing
Neural Network Layers | ✅ Limited | Only a subset of torch.nn layers available
Optimizers | ✅ Limited | Some optimizers (SGD, Adam) exist, but not all
Dataset & Dataloader | 🚫 No | No built-in dataset/dataloader API like torch.utils.data
Distributed Training (DDP) | 🚫 No | No equivalent for torch.distributed
GPU Acceleration (CUDA) | ✅ Yes | CUDA support exists but is less mature than Python's
JIT Compilation | 🚫 No | No TorchScript for model optimization


<hr>
<h2><strong>🔥 What Can You Do with TorchSharp?</strong></h2>
<ul>
<li><strong>Basic Tensor Operations</strong> (e.g., matrix multiplication, broadcasting)</li>
<li><strong>Neural Networks (MLPs, CNNs, RNNs)</strong> using <code inline="">torch.nn</code></li>
<li><strong>Backpropagation &amp; Optimizers</strong> (<code inline="">torch.optim.SGD</code>, <code inline="">torch.optim.Adam</code>)</li>
<li><strong>CUDA Acceleration</strong> (on supported tensors)</li>
</ul>
<h3><strong>✅ Example: Simple TorchSharp Neural Network</strong></h3>
<pre><code class="language-csharp">using System;
using TorchSharp;
using static TorchSharp.torch;
using static TorchSharp.torch.nn;

class Program
{
    static void Main()
    {
        var model = new Sequential(
            Linear(3, 5),
            ReLU(),
            Linear(5, 2),
            Softmax(1)
        );

        var input = torch.randn(4, 3); // Batch of 4 samples, 3 features each
        var output = model.forward(input);

        Console.WriteLine(output);
    }
}
</code></pre>
<h3><strong>📝 What Works in TorchSharp</strong></h3>
<ul>
<li><code inline="">Sequential()</code>, <code inline="">Linear()</code>, <code inline="">ReLU()</code>, and <code inline="">Softmax()</code></li>
<li>Basic forward pass</li>
<li>CPU and CUDA support</li>
</ul>
<hr>
<h2><strong>🔥 When to Use Python’s PyTorch Instead?</strong></h2>
<ul>
<li>If you <strong>need pre-trained models</strong> (e.g., ResNet, BERT)</li>
<li>If you <strong>want advanced deep learning features</strong> (e.g., transformers, diffusion models)</li>
<li>If you <strong>require TorchScript, ONNX, or mobile deployment</strong></li>
<li>If you <strong>need efficient data loading</strong> (TorchSharp lacks <code inline="">DataLoader</code>)</li>
</ul>
<hr>
<h2><strong>🚀 Conclusion</strong></h2>
<p>TorchSharp is great for <strong>basic ML models and tensor operations</strong> in C#, but for <strong>full PyTorch features</strong>, you’ll need to run Python via <strong>Python.NET or an external API</strong>.</p>
<p>Would you like a <strong>workaround to use pretrained models in C#</strong> while still keeping some PyTorch functionality? 🚀</p></body></html>
